codegen/src/main/java/software/amazon/awssdk/codegen/poet/builder/SyncClientBuilderClass.java

```diff
@@ -119,26 +124,46 @@ private MethodSpec endpointProviderMethod() {
-                                       + ".endpointOverride(endpointOverride)"
-                                       + ".build()",
-                                       serviceConfigClassName, serviceConfigClassName, AwsClientOption.class)
-                         .addStatement("return new $T(serviceClientConfiguration, clientConfiguration)", clientClassName)
-                         .build();
+        MethodSpec.Builder b = MethodSpec.methodBuilder("buildClient")
+                                         .addAnnotation(Override.class)
+                                         .addModifiers(Modifier.PROTECTED, Modifier.FINAL)
+                                         .returns(clientInterfaceName)
+                                         .addStatement("$T clientConfiguration = super.syncClientConfiguration()",
+                                                       SdkClientConfiguration.class);
+
+        addQueryProtocolInterceptors(b);
+
+        return b.addStatement("this.validateClientOptions(clientConfiguration)")
+                .addStatement("$T endpointOverride = null", URI.class)
+                .addCode("if (clientConfiguration.option($T.ENDPOINT_OVERRIDDEN) != null"
+                         + "&& $T.TRUE.equals(clientConfiguration.option($T.ENDPOINT_OVERRIDDEN))) {"
+                         + "endpointOverride = clientConfiguration.option($T.ENDPOINT);"
+                         + "}",
+                         SdkClientOption.class, Boolean.class, SdkClientOption.class, SdkClientOption.class)
+                .addStatement("$T serviceClientConfiguration = $T.builder()"
+                              + ".overrideConfiguration(overrideConfiguration())"
+                              + ".region(clientConfiguration.option($T.AWS_REGION))"
+                              + ".endpointOverride(endpointOverride)"
+                              + ".build()",
+                              serviceConfigClassName, serviceConfigClassName, AwsClientOption.class)
+                .addStatement("return new $T(serviceClientConfiguration, clientConfiguration)", clientClassName)
+                .build();
+    }
+
+    private MethodSpec.Builder addQueryProtocolInterceptors(MethodSpec.Builder b) {
+        if (!model.getMetadata().isQueryProtocol()) {
+            return b;
+        }
+
+        TypeName listType = ParameterizedTypeName.get(List.class, ExecutionInterceptor.class);
+        b.addStatement("$T interceptors = clientConfiguration.option($T.EXECUTION_INTERCEPTORS)", listType, SdkClientOption.class)
+            .addStatement("interceptors.add(0, new $T())", QueryParametersToBodyInterceptor.class);
+
+        List<String> customInterceptors = model.getCustomizationConfig().getInterceptors();
+        Collections.reverse(customInterceptors);
+        customInterceptors.forEach(i -> b.addStatement("interceptors.add(0, new $T())", ClassName.bestGuess(i)));
```

Comment:
This is pretty confusing logic. Can we simplify this to remove the reverse and `add(0`?

Reply:
Simplified to use `CollectionUtils.mergeLists()` for clarity